src/Snowflake/Command.php

```diff
@@ -149,18 +226,270 @@ public static function cloneDatabaseFromShared(Connection $connection, array $da
+    public static function createRole(Connection $connection, array $role): void
+    {
+        if (isset($role['granted_by'])) {
+            self::useRole($connection, $role['granted_by']);
+        }
+
+        $connection->query(sprintf(
+            'CREATE ROLE %s',
+            $role['name']
+        ));
+
+        self::grantRoleToUser(
+            $connection,
+            (string) getenv('SNOWFLAKE_DESTINATION_ACCOUNT_USERNAME'),
+            $role['name']
+        );
+    }
+
+    public static function exportGrantsUsersAndRoles(Connection $connection, array $databases): array
+    {
+        $tmp = [];
+        foreach ($databases as $database) {
+            $roles = $connection->fetchAll(sprintf(
+                'SHOW ROLES LIKE %s',
+                QueryBuilder::quote($database . '%')
+            ));
+
+            [
+                'roles' => $roles,
+                'users' => $users,
+            ] = self::assignWorkspacesUsersAndRoles($connection, $roles, [$database]);
+
+            foreach ($users as $user) {
+                $tmp[$database]['users'][] = [
+                    'name' => $user,
+                    'assignedGrants' => $connection->fetchAll(sprintf(
+                        'SHOW GRANTS TO USER %s;',
+                        $user
+                    )),
+                ];
+            }
+
+            foreach ($roles as $role) {
+                $tmp[$database]['roles'][] = array_merge(
+                    $role,
+                    [
+                        'assignedGrants' => $connection->fetchAll(sprintf(
+                            'SHOW GRANTS TO ROLE %s;',
+                            $role['name']
+                        )),
+                        'assignedFutureGrants' => $connection->fetchAll(sprintf(
+                            'SHOW FUTURE GRANTS TO ROLE %s;',
+                            $role['name']
+                        )),
+                    ]
+                );
+            }
+        }
+        return $tmp;
+    }
+
+    private static function assignWorkspacesUsersAndRoles(Connection $connection, array $roles, array $users): array
+    {
+        foreach ($roles as $role) {
+            $grantsToRole = $connection->fetchAll(sprintf(
+                'SHOW GRANTS TO ROLE %s;',
+                $role['name']
+            ));
+
+            $ownershipToRole = array_filter($grantsToRole, fn($v) => $v['privilege'] === 'OWNERSHIP');
+
+            $rolesInRole = array_filter($ownershipToRole, fn($v) => $v['granted_on'] === 'ROLE');
+            $filteredRolesInRole = array_filter($rolesInRole, fn($v) => !in_array($v['name'], $roles));
+            $roles = array_merge($roles, array_combine(
+                array_map(fn($v) => $v['name'], $filteredRolesInRole),
+                $filteredRolesInRole
+            ));
+
+            $usersInRole = array_filter($ownershipToRole, fn($v) => $v['granted_on'] === 'USER');
```

Comment:
V obou případech bych čekal granted_to, pochybuju, že existuje role, která by měla práva něco dělat s uživatelem...

Reply:
tohle je nepoužívaný kus kódu... prověřím a popřípadě smažu.... dělám to jak jsem psal tady - https://github.com/keboola/project-migration-tool/pull/1#discussion_r1087535971

Reply:
> V obou případech bych čekal granted_to, pochybuju, že existuje role, která by měla práva něco dělat s uživatelem...

co tím myslíš? načítám tu jen granty které později použiju

Reply:
"granted on" je "k čemu to může". Takže `$v['granted_on'] === 'USER'` IMHO vyfiltruje granty jako:

`GRANT MONITOR ON USER "KBC_123" TO ROLE "ACCOUNTADMIN";`

Ale nevrátí ti to granty jako:
`GRANT USAGE ON ROLE "KBC_123_RO" TO USER "TOMAS_FEJFAR";`

Myslím, že tam chceš "grantee name" nebo "granted to" (nevím, jak přesně se to jmenuje). Prostě si myslím, že v tom filtru koukáš na špatný sloupeček...

Reply:
nakonec jsem ty uživatele odsud úplně smazal a udělal to úplně jinak